### PR TITLE
Improve questions, add question sets, and add ability to switch between question sets.

### PR DESCRIPTION
**Pull request by eyurko**
I think I got most of the things I've noticed.

For a variety of trying-not-to-shoot-myself-in-the-foot-long-term reasons, this also includes the "Add ability to change question sets" PR (hate them merge conflicts), so I'm closing #238.

### CODE REVIEW

**Comment by eyurko** [comment]
Would like CSS / front-end approval from @theresa-lee, as well.

File: client/components/LearnerViewDirective.js [hunk 1]
```diff
@@ -80,6 +80,12 @@ tie.directive('learnerView', [function() {
                   <select class="tie-lang-select-menu" name="lang-select-menu">
                     <option value="Python" selected>Python</option>
                   </select>
+                  <select class="tie-question-set-select" name="question-set-select"
+                          ng-change="changeQuestionSet()" ng-model="currentQuestionSetId"
+                          ng-options="i.questionSetId as i.questionSetId for i in questionSetIds">
+                    <option style="display: none" value="">Question Set</option>
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Why do you need this and the next inner option?

File: client/components/LearnerViewDirective.js [hunk 2]
```diff
@@ -450,6 +462,12 @@ tie.directive('learnerView', [function() {
         // TODO(sll): Generalize this to dynamically select a question set
         // based on user input.
         var questionSetId = 'strings';
+        var questionSetIds = [
+          {questionSetId: 'strings'},
+          {questionSetId: 'all'},
+          {questionSetId: 'other'}
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Perhaps "other" before "all"? "all" seems like it's just for testing.

File: client/components/LearnerViewDirective.js [hunk 3]
```diff
@@ -568,6 +586,33 @@ tie.directive('learnerView', [function() {
           }
         };
 
+        $scope.changeQuestionSet = function() {
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: I suggest actually passing in the questionSetId as an argument.

File: client/components/LearnerViewDirective.js [hunk 4]
```diff
@@ -568,6 +586,33 @@ tie.directive('learnerView', [function() {
...
+          questionSetId = '';
+          switch ($scope.currentQuestionSetId) {
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Cleaner to have, somewhere:

    var ALLOWED_QUESTION_SET_IDS = [...];

and then do

    if (ALLOWED_QUESTION_SET_IDS.indexOf(questionSetId) === -1) {
      return;
    }

    ...

File: client/components/LearnerViewDirective.js [hunk 5]
```diff
@@ -568,6 +586,33 @@ tie.directive('learnerView', [function() {
...
+          case 'all':
+            questionSetId = $scope.currentQuestionSetId;
+            break;
+          default:
+            return;
+          }
+          QuestionDataService.initCurrentQuestionSet(questionSetId);
+          questionSet = QuestionDataService.getCurrentQuestionSet(
+            questionSetId);
+          $scope.currentQuestionIndex = 0;
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Don't copy/paste. Extract all this setup stuff into another method (initQuestionSet, perhaps).

File: client/components/LearnerViewDirective.js (hunk 1 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: I think that one's the title, but I don't think I need the next one.

Comment (eyurko) [inline client/components/LearnerViewDirective.js]: Yeah -- that's done for consistency with the "Theme" dropdown, below. It also has the <option></option>, would you like me to remove that one as well?

File: client/components/LearnerViewDirective.js (hunk 3 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: This is also consistent with the implementation of the themes -- should we change the way we do theming, as well?

File: client/components/LearnerViewDirective.js (hunk 1 above)
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Yeah, let's remove stuff that's not needed. Thanks!

File: client/components/LearnerViewDirective.js (hunk 3 above)
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Yes, that would be better, I think, if that's OK!

File: client/components/LearnerViewDirective.js (hunk 5 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: done

File: client/components/LearnerViewDirective.js (hunk 4 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: done, though I will miss the switch statements.

File: client/components/LearnerViewDirective.js (hunk 2 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: sure, sounds good.

File: client/components/LearnerViewDirective.js (hunk 3 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: aight

File: client/components/LearnerViewDirective.js (hunk 1 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: aight

File: client/components/LearnerViewDirective.js [hunk 6]
```diff
@@ -557,17 +562,42 @@ tie.directive('learnerView', [function() {
           $scope.$apply();
         };
 
-        $scope.changeTheme = function() {
-          if ($scope.theme === 'Dark') {
+        $scope.changeTheme = function(theme) {
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Suggest newTheme, for parity with newQuestionSetId.

File: client/components/LearnerViewDirective.js [hunk 7]
```diff
@@ -446,10 +456,17 @@ tie.directive('learnerView', [function() {
           CodeStorageService, SECONDS_TO_MILLISECONDS, DEFAULT_AUTOSAVE_SECONDS,
           DISPLAY_AUTOSAVE_TEXT_SECONDS) {
         var DURATION_MSEC_WAIT_FOR_SCROLL = 20;
+        var ALLOWED_QUESTION_SET_IDS = ['strings', 'other', 'all'];
         var language = LANGUAGE_PYTHON;
         // TODO(sll): Generalize this to dynamically select a question set
         // based on user input.
         var questionSetId = 'strings';
+        var questionSetIds = [
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Not sure you need the "var" variable -- you can just define $scope.questionSetIds directly.

That said, probably better not to hardcode two things with the same data. Can $scope.questionSetIds be derived from ALLOWED_QUESTION_SET_IDS somehow?

File: client/components/LearnerViewDirective.js (hunk 6 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: done

File: client/components/LearnerViewDirective.js (hunk 7 above)
Comment (eyurko) [inline client/components/LearnerViewDirective.js]: done

**Comment by seanlip** [comment]
LGTM, though I think you still want @theresa-lee's approval. Thanks @eyurko!

**Pull request merged**